controllers/networkfence_controller.go

```diff
@@ -113,37 +118,68 @@ func (r *NetworkFenceReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, err
 	}
 
-	// check if the networkfence object is getting deleted, if yes, then send
-	// UnfenceCluster request to the driver.
-	// After the UnfenceCluster request is processed, the finalizer will be removed
-	// so that the networkfence object can be deleted gracefully.
-	if !nwFence.GetDeletionTimestamp().IsZero() {
+	nf := NetworkFenceInstance{
+		ctx:              ctx,
+		reconciler:       r,
+		logger:           logger,
+		instance:         nwFence,
+		controllerClient: client,
+	}
+
+	// check if the networkfence object is getting deleted and handle it.
+	if !nf.instance.GetDeletionTimestamp().IsZero() {
 		if util.ContainsInSlice(nwFence.GetFinalizers(), networkFenceFinalizer) {
 
-			err := r.unfenceClusterNetwork(ctx, nwFence, client, logger)
+			err := nf.removeFinalizerFromNetworkFence()
 			if err != nil {
-				logger.Error(err, "failed to unfence cluster network")
+				logger.Error(err, "failed to remove finalizer on network fence resource")
 				return ctrl.Result{}, err
 			}
 		}
+
 		logger.Info("NetworkFence object is terminated, skipping reconciliation")
 		return ctrl.Result{}, nil
 	}
 
-	err = r.fenceClusterNetwork(ctx, nwFence, client, logger)
+	err = nf.processFencing()
 	if err != nil {
 		logger.Error(err, "failed to fence cluster network")
+		updateStatusErr := nf.updateStatus(csiaddonsv1alpha1.FencingOperationResultFailed, "fencing operation failed")
+		if updateStatusErr != nil {
+			logger.Error(updateStatusErr, "failed to update status")
+		}
+
 		return ctrl.Result{}, err
 	}
-	nwFence.Status.Result = csiaddonsv1alpha1.FencingOperationResultSucceeded
-	nwFence.Status.Message = "NetworkFence operation succeeded"
-	if err := r.Client.Status().Update(ctx, nwFence); err != nil {
+
+	if err = nf.updateStatus(csiaddonsv1alpha1.FencingOperationResultSucceeded, "fencing operation successful"); err != nil {
+		logger.Error(err, "failed to update status")
 		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
+type NetworkFenceInstance struct {
```

Comment:
Missing documentation for this type and its attributes.

Storing `context.Context` in a struct is not common, it usually gets passed as an argument to functions.

Reply:
Will be adding the doc for this type.

This struct is used here as a kind of class where the attributes are the members of it and there are methods (such as fenceClusterNetwork, unfenceClusterNetwork) that need the attributes from the struct "NetworkFenceInstance". And this helped in keeping the functions simple. Also, the struct NetworkFenceInstace has not been passed as an argument to any function. 

This is based on some of the controller implementations in the ramen project which simplified the implementation.

https://github.com/RamenDR/ramen/blob/main/controllers/drplacementcontrol.go#L43
https://github.com/RamenDR/ramen/blob/main/controllers/volumereplicationgroup_controller.go#L316

However, I am fine with removing the attribute "ctx context.Context" from the struct "NetworkFenceInstance" and instead pass "ctx context.Context" as the argument to the functions where necessary.

Please let me know your thoughts.

Reply:
Ok, I think that is fine. Thanks for the explanation.